### PR TITLE
lib: pcm_mix: Moved PCM mix from application

### PR DESCRIPTION
**Pull request by koffes**
Tests and code moved from applications: nrf5340_audio

Signed-off-by: Kristoffer Rist Skøien <kristoffer.skoien@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/libraries/others/pcm_mix.rst [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+The PCM mixer enables users to mix a PCM stream into another PCM stream.
+There is support for mixing a stereo stream into another stereo stream,
+or, combinations of mono to mono, or, mono to stereo channel left or right.
+
+Configuration
+*************
+
+To enable the library, set the :kconfig:option:`CONFIG_PCM_MIX` Kconfig option to ``y`` in the project configuration file :file:`prj.conf`.
```
Comment (koffes) [inline doc/nrf/libraries/others/pcm_mix.rst:18]: Need to confirm is this is the preferred method.

**Review by rick1082 — APPROVED** [review]

File: tests/lib/pcm_mix/src/main.c [hunk 2]
```diff
@@ -23,6 +23,9 @@ void test_mix_with_nothing(void)
 	int16_t sample_a[] = { 0, 1, -1 };
 	int16_t sample_r[] = { 0, 1, -1 };
 
+	/*TODO: injected failure */
+	assert_equal(1, 0, "Failing on purpose");
```
Comment (alexsven) [inline tests/lib/pcm_mix/src/main.c:27]: Remove before merging?

**Review by alexsven — APPROVED** [review]

File: doc/nrf/libraries/others/pcm_mix.rst [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+   :depth: 2
+
+The Pulse Code Modulation (PCM) audio mixer lets you mix a PCM stream into another PCM stream.
+This library is useful for developing applications that offer audio features, for example using the nRF5340 Audio DK.
+
+The library offers the following types of mixing support:
+
+* Stereo stream into another stereo stream
+* Combinations of mono to mono
+* Mono to stereo channel left or right
```
Comment (greg-fer) [inline doc/nrf/libraries/others/pcm_mix.rst:17]: Is "channel left or right" referring to "stereo" only or to both "mono and stereo"?

File: doc/nrf/libraries/others/pcm_mix.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+Pulse Code Modulation audio mixer
+#################################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The Pulse Code Modulation (PCM) audio mixer lets you mix a PCM stream into another PCM stream.
+This library is useful for developing applications that offer audio features, for example using the nRF5340 Audio DK.
```
Comment (greg-fer) [inline doc/nrf/libraries/others/pcm_mix.rst:11]: @koffes , we need a use case/purpose for the library. This is a placeholder, please provide a better one if you think it doesn't work.

Comment (erikrobstad) [reply doc/nrf/libraries/others/pcm_mix.rst:11]: Maybe something like:

"by for example mixing a confirmation tone together with the music that is playing"

This is what we are using pcm_mix for today in the nrf5340_audio application

File: doc/nrf/libraries/others/pcm_mix.rst (hunk 1 above)
Comment (erikrobstad) [reply doc/nrf/libraries/others/pcm_mix.rst:22]: What do you mean by this @koffes ?

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: include/pcm_mix.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * @retval -EINVAL      pcm_a is NULL or size_a = 0.
+ * @retval -EPERM       Either size_b < size_a (for stereo to stereo, mono to mono)
+ *			or size_a/2 < size_b (for mono to stereo mix).
+ * @retval -ESRCH       Invalid mixing mode.
+ */
+int pcm_mix(void *const pcm_a, size_t size_a, void const *const pcm_b, size_t size_b,
+	    enum pcm_mix_mode mix_mode);
+
+/**
+ * @}
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply include/pcm_mix.h:44]: Yes, for the API doc to generate correctly.

**Review by greg-fer — CHANGES_REQUESTED** [review]
Please make sure to check the doxygen guidelines: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#doxygen-gl

File: include/pcm_mix.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply include/pcm_mix.h:44]: Added the opening doxygen lines in the request changes comments.

File: doc/nrf/libraries/others/pcm_mix.rst (hunk 4 above)
Comment (koffes) [reply doc/nrf/libraries/others/pcm_mix.rst:11]: Changed

File: doc/nrf/libraries/others/pcm_mix.rst (hunk 3 above)
Comment (koffes) [reply doc/nrf/libraries/others/pcm_mix.rst:17]: I don't quite understand. You can mix a mono stream to stereo-L or stereo-R or stereo L+R.
See updated version which I will push before lunch.

File: doc/nrf/libraries/others/pcm_mix.rst (hunk 1 above)
Comment (koffes) [reply doc/nrf/libraries/others/pcm_mix.rst:22]: Have a look now.

File: tests/lib/pcm_mix/src/main.c (hunk 2 above)
Comment (koffes) [reply tests/lib/pcm_mix/src/main.c:27]: Yes. I will do that as soon that I verify that this fails.

File: lib/pcm_mix/Kconfig [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# PCM (Pulse Code Modulation) audio mixer
```
Comment (erikrobstad) [inline lib/pcm_mix/Kconfig:1]: Nitpick: Does not look like the other Kconfig files have this on the first line

File: doc/nrf/libraries/others/pcm_mix.rst (hunk 1 above)
Comment (erikrobstad) [reply doc/nrf/libraries/others/pcm_mix.rst:24]: I meant the sentence "Need to confirm is this is the preferred method.", but looks like this is good now :)

File: include/pcm_mix.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/** @file
```
Comment (erikrobstad) [inline include/pcm_mix.h:7]: Nitpick:

```suggestion
/** 
  *@file
```

File: tests/lib/pcm_mix/src/main.c (hunk 2 above)
Comment (erikrobstad) [reply tests/lib/pcm_mix/src/main.c:27]: This still needs to be removed :)

**Review by erikrobstad — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Pull request merged**